### PR TITLE
mon: show no[deep-]scrub flags per pool in the status

### PR DESCRIPTION
**Pull request by mogeb**
Fixes: http://tracker.ceph.com/issues/38029

Signed-off-by: Mohamad Gebai <mgebai@suse.com>

- [x] References tracker ticket
- [ ] Updates documentation if necessary
- [x] Includes tests for new functionality or reproducer for bug

### CODE REVIEW

**Comment by mogeb** [comment]
If no pools have any no*scrub flags, the status looks like:

```
$> /bin/ceph -s
  cluster:
    id:     e8f8a0a9-75d9-4097-a9db-9b56044d1d14
    health: HEALTH_OK
 
  services:
    mon: 1 daemons, quorum a (age 68s)
    mgr: x(active, since 59s)
    osd: 3 osds: 3 up (since 21s), 3 in (since 21s)
 
  data:
    pools:   0 pools, 0 pgs
    objects: 0 objects, 0 B
    usage:   3.2 GiB used, 27 GiB / 30 GiB avail
    pgs:     
 
  notes:
 ```

With per-pool flags:
```
$> ./bin/ceph -s
  cluster:
    id:     e8f8a0a9-75d9-4097-a9db-9b56044d1d14
    health: HEALTH_OK
 
  services:
    mon: 1 daemons, quorum a (age 26s)
    mgr: x(active, since 13m)
    osd: 3 osds: 3 up (since 12m), 3 in (since 12m)
 
  data:
    pools:   2 pools, 140 pgs
    objects: 0 objects, 0 B
    usage:   3.2 GiB used, 27 GiB / 30 GiB avail
    pgs:     140 active+clean
 
  notes:
    pool rbd has noscrub flag 
    pool rbd has nodeep-scrub flag 
    pool mypool has nodeep-scrub flag 
```
With the JSON formatter:
```
$> ./bin/ceph -s --format json | jq .notes
{
  "pool rbd flags": "noscrub nodeep-scrub",
  "pool mypool flags": " nodeep-scrub"
}
```

Thoughts on the output? I'm not really fond of the JSON one.

**Comment by jecluis** [comment]
> With the JSON formatter:
> 
> ```
> $> ./bin/ceph -s --format json | jq .notes
> {
>   "pool rbd flags": "noscrub nodeep-scrub",
>   "pool mypool flags": " nodeep-scrub"
> }
> ```
> Thoughts on the output? I'm not really fond of the JSON one. It's also misleading as one might think that only these scrub flags are enabled for a pool, whereas a pool might have more flags enabled.

Yeah, neither am I. I think this should be presented, as JSON, a bit like the health status we use for the checks. Something like
```
{ "notes": [
    "POOL_FLAG_INFO": {
        "pool": "rbd",
        "flags:" [ "noscrub", "nodeep-scrub"]
    },
    "POOL_FLAG_INFO": {
        "pool": "mypool",
        "flags:" ["nodeep-scrub"]
    }]
}
```
or some such (not married to the `POOL_FLAG_INFO` thing either, but couldn't come up with anything else).

Additionally, I don't like the `notes` either. I think this should actually go into the health checks, just not cause a WARN or an ERR. We could, possibly, show that info under `pools` (in `data`) when using the plain-text output though.

**Comment by mogeb** [comment]
> ```
> { "notes": [
>     "POOL_FLAG_INFO": {
>         "pool": "rbd",
>         "flags:" [ "noscrub", "nodeep-scrub"]
>     },
>     "POOL_FLAG_INFO": {
>         "pool": "mypool",
>         "flags:" ["nodeep-scrub"]
>     }]
> }
> ```

It does look better. If we go that way, I suggest `POOL_SCRUB_INFO` instead of `POOL_FLAG_INFO`, since we'll only show *scrub* flags, and not *all* flags.

> Additionally, I don't like the `notes` either. I think this should actually go into the health checks, just not cause a WARN or an ERR. We could, possibly, show that info under `pools` (in `data`) when using the plain-text output though.

Yep I agree. I'll try that.

**Comment by trociny** [comment]
I am not sure `ceph status` is the right place for information like this. I would like to keep it laconic as much as possible. IMHO what would probably be ok is to display info (no warning) that scrub/deep-scrub is disabled for some pools, without specifying for which ones (could be easily found with other commands). And only if it would not make the output look ugly or too much verbose.

Again, just IMHO.

**Comment by jecluis** [comment]
@mogeb any update here?

**Comment by mogeb** [comment]
@jecluis do you have any thought on @trociny's comments? There's a point to keeping the `ceph -s` command's output concise.

**Comment by jecluis** [comment]
yeah, maybe just making sure it shows on detail would be enough.

**Comment by mogeb** [comment]
@jecluis  Update 1:
```
$> ./bin/ceph -s
  cluster:
    id:     44313e9f-0433-42a3-84d6-c57019c94c9f
    health: HEALTH_OK
            Some pool(s) have the noscrub, nodeep-scrub flag(s) set
 
  services:
    mon: 1 daemons, quorum a (age 34s)
    mgr: x(active, since 31s)
    osd: 1 osds: 1 up (since 17s), 1 in (since 17s)
 
  data:
    pools:   2 pools, 64 pgs
    objects: 0 objects, 0 B
    usage:   1.1 GiB used, 9.0 GiB / 10 GiB avail
    pgs:     64 active+clean
```

```json
$> ./bin/ceph -s --format=json | jq .health
{
  "checks": {
    "POOL_SCRUB_FLAGS": {
      "severity": "HEALTH_OK",
      "summary": {
        "message": "Some pool(s) have the noscrub, nodeep-scrub flag(s) set"
      }
    }
  },
  "status": "HEALTH_OK"
}
```

```
$> ./bin/ceph report | jq .health
{
  "checks": {
    "POOL_SCRUB_FLAGS": {
      "severity": "HEALTH_OK",
      "summary": {
        "message": "Some pool(s) have the noscrub, nodeep-scrub flag(s) set"
      },
      "detail": [
        {
          "message": "Pool rbd has noscrub flag"
        },
        {
          "message": "Pool mypool has noscrub flag"
        },
        {
          "message": "Pool mypool has nodeep-scrub flag"
        }
      ]
    }
  },
  "status": "HEALTH_OK"
}
```

```
$> ./bin/ceph health
HEALTH_OK Some pool(s) have the noscrub, nodeep-scrub flag(s) set
```

```
$> ./bin/ceph health detail
HEALTH_OK Some pool(s) have the noscrub, nodeep-scrub flag(s) set
POOL_SCRUB_FLAGS Some pool(s) have the noscrub, nodeep-scrub flag(s) set
    Pool rbd has noscrub flag
    Pool mypool has noscrub flag
    Pool mypool has nodeep-scrub flag
```

File: src/osd/OSDMap.cc [hunk 1]
```diff
@@ -5157,6 +5157,21 @@ void OSDMap::check_health(health_check_map_t *checks) const
     }
   }
 
+  auto& d = checks->add("POOL_SCRUB_FLAGS", HEALTH_OK, "Some pool(s) have the noscrub, nodeep-scrub flag(s) set");
```
Comment (jecluis) [inline src/osd/OSDMap.cc]: pretty sure this needs to be wrapped.

Comment (jecluis) [reply src/osd/OSDMap.cc]: also, would be nice if we have the message providing whether it's just `noscrub` or `nodeep-scrub`, or both.

Comment (mogeb) [reply src/osd/OSDMap.cc]: Oops, of course it needs to wrapped. Fixed. Per your second comment, I made it a bit better, but we can't really know if pools have one flag, or the other, or both. I think it's acceptable though as one can easily find it using `ceph health detail`.

**Comment by jecluis** [comment]
@mogeb looks good to me. Can you please add a test to `cephtool/test.sh`?

**Comment by mogeb** [comment]
@jecluis done.

**Comment by mogeb** [comment]
@jecluis anything blocking this?

**Comment by jecluis** [comment]
@mogeb I'm very tempted to block it on the account of your test going well beyond the 80 cols line length ;)

**Comment by mogeb** [comment]
@jecluis hehe oops! fixed :)

**Comment by mogeb** [comment]
@jecluis anything else blocking this?

**Pull request merged**